Sources/Plasma/CoreLib/HeadSpin_Win.cpp

```diff
@@ -0,0 +1,219 @@
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "HeadSpin.h"
+#include "hsWindows.h"
+
+#include <string_theory/string>
+
+class hsMinimizeClientGuard
+{
+    hsWindowHndl fWnd;
+
+public:
+    hsMinimizeClientGuard()
+    {
+        fWnd = GetActiveWindow();
+        // If the application's topmost window is fullscreen, minimize it before displaying an error
+        if ((GetWindowLong(fWnd, GWL_STYLE) & WS_POPUP) != 0)
+            ShowWindow(fWnd, SW_MINIMIZE);
+    }
+
+    ~hsMinimizeClientGuard()
+    {
+        ShowWindow(fWnd, SW_RESTORE);
+    }
+};
+
+
+int hsMessageBoxWithOwner(hsWindowHndl owner, const ST::string& message, const ST::string& caption, int kind, int icon)
+{
+    if (hsMessageBox_SuppressPrompts)
+        return hsMBoxOk;
+
+    uint32_t flags = 0;
+
+    if (kind == hsMessageBoxNormal)
+        flags |= MB_OK;
+    else if (kind == hsMessageBoxAbortRetyIgnore)
+        flags |= MB_ABORTRETRYIGNORE;
+    else if (kind == hsMessageBoxOkCancel)
+        flags |= MB_OKCANCEL;
+    else if (kind == hsMessageBoxRetryCancel)
+        flags |= MB_RETRYCANCEL;
+    else if (kind == hsMessageBoxYesNo)
+        flags |= MB_YESNO;
+    else if (kind == hsMessageBoxYesNoCancel)
+        flags |= MB_YESNOCANCEL;
+    else
+        flags |= MB_OK;
+
+    if (icon == hsMessageBoxIconError)
+        flags |= MB_ICONERROR;
+    else if (icon == hsMessageBoxIconQuestion)
+        flags |= MB_ICONQUESTION;
+    else if (icon == hsMessageBoxIconExclamation)
+        flags |= MB_ICONEXCLAMATION;
+    else if (icon == hsMessageBoxIconAsterisk)
+        flags |= MB_ICONASTERISK;
+    else
+        flags |= MB_ICONERROR;
+
+    hsMinimizeClientGuard guard;
+    int ans = MessageBoxW(owner, message.to_wchar().data(), caption.to_wchar().data(), flags);
+
+    switch (ans)
+    {
+    case IDOK:          return hsMBoxOk;
+    case IDCANCEL:      return hsMBoxCancel;
+    case IDABORT:       return hsMBoxAbort;
+    case IDRETRY:       return hsMBoxRetry;
+    case IDIGNORE:      return hsMBoxIgnore;
+    case IDYES:         return hsMBoxYes;
+    case IDNO:          return hsMBoxNo;
+    default:            return hsMBoxCancel;
+    }
+
+    return hsMBoxCancel;
+}
+
+int hsMessageBoxWithOwner(hsWindowHndl owner, const char* message, const char* caption, int kind, int icon)
```

Comment:
Why not just pass these variants onto the `ST::string` variant, and then eventually deprecate/remove the non-`ST::string` overloads?  Otherwise, this is a lot of duplication...

Reply:
That's what I was thinking, and performance of string conversions probably doesn't matter much in the case of message boxes